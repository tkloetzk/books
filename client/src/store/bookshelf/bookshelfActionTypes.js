export const FETCH_BOOKSHELF_HAS_ERRORED = 'FETCH_BOOKSHELF_HAS_ERRORED';
export const FETCH_BOOKSHELF_IS_LOADING = 'FETCH_BOOKSHELF_IS_LOADING';
export const FETCH_BOOKSHELF_SUCCESS = 'FETCH_BOOKSHELF_SUCCESS';
export const FILTER_BOOKSHELF_SUCCESS = 'FILTER_BOOKSHELF_SUCCESS';
export const SAVE_COMBINED_BOOKS_SUCCESS = 'SAVE_COMBINED_BOOKS_SUCCESS';
export const ADD_BOOK_TO_BOOKSHELF_SUCCESS = 'ADD_BOOK_TO_BOOKSHELF_SUCCESS';
export const ADD_BOOK_TO_BOOKSHELF_FAILURE = 'ADD_BOOK_TO_BOOKSHELF_FAILURE';
export const UPDATE_BOOK_ON_BOOKSHELF_SUCCESS =
  'UPDATE_BOOK_ON_BOOKSHELF_SUCCESS';
export const UPDATE_BOOK_ON_BOOKSHELF_FAILURE =
  'UPDATE_BOOK_ON_BOOKSHELF_FAILURE';
export const SAVE_MODIFIED_BOOKS_SUCCESS = 'SAVE_MODIFIED_BOOKS_SUCCESS';
export const UPDATE_MODIFIED_BOOK_SUCCESS = 'UPDATE_MODIFIED_BOOK_SUCCESS';
export const INSERT_MODIFIED_BOOK_SUCCESS = 'INSERT_MODIFIED_BOOK_SUCCESS';
export const DELETE_BOOK_ON_BOOKSHELF_FAILURE =
  'DELETE_BOOK_ON_BOOKSHELF_FAILURE';
export const DELETE_BOOK_ON_BOOKSHELF_SUCCESS =
  'DELETE_BOOK_ON_BOOKSHELF_SUCCESS';
export const DELETE_BOOK_FROM_BOOKLIST_SUCCESS =
  'DELETE_BOOK_FROM_BOOKLIST_SUCCESS';
export const FETCH_BOOKSHELF_GENRES_SUCCESS = 'FETCH_BOOKSHELF_GENRES_SUCCESS';
export const FETCH_BOOKSHELF_GENRES_FAILURE = 'FETCH_BOOKSHELF_GENRES_FAILURE';
export const SELECTED_GENRES = 'SELECTED_GENRES';
