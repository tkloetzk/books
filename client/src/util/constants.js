export const LOADING_STATUSES = {
  initial: 'initial',
  loading: 'loading',
  success: 'success',
  errored: 'errored',
};
