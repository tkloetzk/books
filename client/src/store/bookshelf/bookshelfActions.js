import get from 'lodash/get';
import {
  addBookshelfService,
  getBookshelfService,
} from '../../services/bookshelfService';
import * as types from './bookshelfActionTypes';

//TODO: probably shouldn't be modifying original booklist
export default function sortBooklist(booklist) {
  calculateMeans(booklist);
  calculateAdjustedRating(booklist);
  booklist.sort((a, b) =>
    a.adjustedRating < b.adjustedRating
      ? 1
      : b.adjustedRating < a.adjustedRating
      ? -1
      : 0
  );
  return booklist;
}

function calculateMeans(booklist) {
  let goodreadsVotes = 0,
    amazonVotes = 0,
    total = 0;
  for (var i = 0; i < booklist.length; i++) {
    //TODO: Best for loop?
    // for (Book book: booklist) { // TODO Iterator
    goodreadsVotes += get(booklist[i], 'goodreadsAverageRating');
    amazonVotes += get(booklist[i], 'amazonAverageRating');
    total +=
      get(booklist[i], 'goodreadsAverageRating') +
      get(booklist[i], 'amazonAverageRating');
  }

  booklist.meanGoodreadsVotes = goodreadsVotes / booklist.length;
  booklist.meanAmazonVotes = amazonVotes / booklist.length;
  booklist.totalMean = total / 2 / booklist.length;
  return booklist;
}

function calculateAdjustedRating(booklist) {
  for (var i = 0; i < booklist.length; i++) {
    var book = booklist[i];
    var adjustedRating =
      (getAdjustedRating(
        get(book, 'goodreadsRatingsCount'),
        get(book, 'goodreadsAverageRating'),
        get(booklist, 'meanGoodreadsVotes'),
        trimmean(getGoodreadsRatingsCountList(booklist))
      ) +
        getAdjustedRating(
          get(book, 'amazonRatingsCount'),
          get(book, 'amazonAverageRating'),
          get(booklist, 'meanAmazonVotes'),
          trimmean(getAmazonRatingsCountList(booklist))
        )) /
      2;
    book.adjustedRating = adjustedRating;
  }
}

function getAdjustedRating(ratingsCount, averageRating, meanVote, minVotes) {
  //double rating = averageRating * 2;
  return (
    (ratingsCount / (ratingsCount + minVotes)) * averageRating +
    (minVotes / (ratingsCount + minVotes)) * meanVote
  );
}

function getGoodreadsRatingsCountList(booklist) {
  let votes = [];
  for (const book of booklist) {
    votes.push(book.goodreadsRatingsCount);
  }

  return votes.sort(function(a, b) {
    return a - b;
  });
}

function getAmazonRatingsCountList(booklist) {
  let votes = [];
  for (const book of booklist) {
    votes.push(book.amazonRatingsCount);
  }

  return votes.sort(function(a, b) {
    return a - b;
  });
}

function trimmean(votes) {
  votes.sort(function(a, b) {
    return a - b;
  });

  const removeAmount = (votes.length * 0.25) / 2;
  const roundDownNearedMultipleTwo = Math.floor(
    removeAmount >= 0
      ? (removeAmount / 2) * 2
      : ((removeAmount - 2 + 1) / 2) * 2
  );
  return getMean(
    votes.slice(
      roundDownNearedMultipleTwo,
      votes.length - roundDownNearedMultipleTwo
    )
  );
}

function getMean(array) {
  let sum = 0;
  for (let i = 0; i < array.length; i++) {
    sum += array[i];
  }
  return sum / array.length;
}

export function getBookshelfIsLoading(bool) {
  return {
    type: types.FETCH_BOOKSHELF_IS_LOADING,
    isLoading: bool,
  };
}

export function getBookshelfSuccess(bookshelf) {
  return {
    type: types.FETCH_BOOKSHELF_SUCCESS,
    bookshelf,
  };
}

export function getBookshelfFailure(bool) {
  return {
    type: types.FETCH_BOOKSHELF_HAS_ERRORED,
    hasErrored: bool,
  };
}

export function saveCombinedBooksSuccess(booklist) {
  return {
    type: types.SAVE_COMBINED_BOOKS_SUCESS,
    booklist,
  };
}
export function saveCombinedBooks(books) {
  return dispatch => {
    dispatch(saveCombinedBooksSuccess(books));
  };
}
export function getBookshelf(excludeGenre = []) {
  return dispatch => {
    dispatch(getBookshelfIsLoading(true));
    return getBookshelfService(excludeGenre)
      .then(bookshelf => {
        dispatch(getBookshelfIsLoading(false));
        dispatch(getBookshelfSuccess(bookshelf));
        return bookshelf;
      })
      .catch(error => {
        dispatch(getBookshelfIsLoading(false));
        dispatch(getBookshelfFailure(true));
        console.error('bookshelf error', error);
      });
  };
}

export function addBookToBookshelfSuccess(booklist) {
  return {
    type: types.ADD_BOOK_TO_BOOKSHELF_SUCCESS,
    booklist,
  };
}
export function addBookToBookshelf(booklist) {
  return dispatch => {
    return addBookshelfService(booklist)
      .then(resp => {
        dispatch(addBookToBookshelfSuccess(booklist));
        return true;
      })
      .catch(error => {
        console.log('error', error);
      });
  };
}
