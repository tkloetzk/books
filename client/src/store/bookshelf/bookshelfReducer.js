import * as types from './bookshelfActionTypes';
import { LOADING_STATUSES } from '../../util/constants';
import some from 'lodash/some';
import remove from 'lodash/remove';

const initialState = {
  hasErrored: null,
  bookshelf: [],
  booklist: [],
  modifiedBooklist: [],
  error: null,
  saveStatus: { status: LOADING_STATUSES.initial, message: '' },
  refreshed: false,
};

export default function bookshelf(state = initialState, action) {
  switch (action.type) {
    case types.FETCH_BOOKSHELF_SUCCESS:
      return Object.assign({}, state, {
        bookshelf: action.bookshelf,
      });
    case types.FETCH_BOOKSHELF_HAS_ERRORED:
      return Object.assign({}, state, {
        hasErrored: action.hasErrored,
      });
    case types.SAVE_COMBINED_BOOKS_SUCCESS:
      return Object.assign({}, state, {
        booklist: [...action.booklist], // TODO: Do i need to array and deconstructing?
      });
    //TODO: Could save and insert be combined?
    case types.SAVE_MODIFIED_BOOKS_SUCCESS:
      return Object.assign({}, state, {
        modifiedBooklist: action.modifiedBooklist,
      });
    case types.INSERT_MODIFIED_BOOK_SUCCESS:
      const { modifiedBook } = action;
      const { modifiedBooklist } = state;

      var exisiting = some(
        modifiedBooklist,
        book => book.isbn === modifiedBook.isbn
      );

      if (exisiting) {
        const newModifiedBooklist = modifiedBooklist.map(book => {
          if (book.isbn === modifiedBook.isbn)
            return Object.assign({}, book, modifiedBook);
          return book;
        });
        return Object.assign({}, state, {
          modifiedBooklist: newModifiedBooklist,
        });
      }
      remove(state.booklist, book => book.isbn === modifiedBook.isbn); // TODO: What is this doing?
      return Object.assign({}, state, {
        modifiedBooklist: [...state.modifiedBooklist, modifiedBook],
      });
    case types.ADD_BOOK_TO_BOOKSHELF_SUCCESS: {
      return Object.assign({}, state, {
        booklist: action.booklist,
        saveStatus: {
          status: LOADING_STATUSES.success,
          message: 'Save Successful',
        },
      });
    }
    case types.UPDATE_BOOK_ON_BOOKSHELF_SUCCESS: {
      return Object.assign({}, state, {
        modifiedBooklist: action.modifiedBooklist,
        saveStatus: {
          status: LOADING_STATUSES.success,
          message: 'Save Successful',
        },
      });
    }
    case types.ADD_BOOK_TO_BOOKSHELF_FAILURE: {
      return Object.assign({}, state, {
        saveStatus: {
          status: LOADING_STATUSES.errored,
          message: 'Save Failed',
        },
      });
    }
    case types.REFRESHED_BOOKSHELF: {
      return Object.assign({}, state, {
        refreshed: action.refreshed,
      });
    }
    default:
      return state;
  }
}
