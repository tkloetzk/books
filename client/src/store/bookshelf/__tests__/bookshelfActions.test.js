import * as actions from '../bookshelfActions';
import * as types from '../bookshelfActionTypes';
import { LOADING_STATUSES } from '../../../util/constants';
import {
  addBookshelfService,
  getBookshelfService,
  deleteBookOnBookshelfService,
  updateBookOnBookshelfService,
  getGenresBookshelfService,
} from '../../../services/bookshelfService';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { initialState } from '../bookshelfReducer';

const createMockStore = configureMockStore([thunk]);

jest.mock('../../../services/bookshelfService');

describe('bookshelf actions', () => {
  let booklist;
  beforeEach(() => {
    booklist = [
      {
        amazonAverageRating: 4.6,
        amazonRatingsCount: 29,
        price: '',
        isbn: '9780736917728',
        title: "Raising a Daughter After God's Own Heart",
        subtitle: '',
        description:
          'Elizabeth George, bestselling author and mother of two daughters',
        thumbnail:
          'http://books.google.com/books/content?id=K_AhmQEACAAJ&printsec=frontcover',
        categories: ['Religion'],
        goodreadsAverageRating: 4.25,
        goodreadsRatingsCount: 149,
      },
    ];

    addBookshelfService.mockClear();
    getBookshelfService.mockClear();
    deleteBookOnBookshelfService.mockClear();
    updateBookOnBookshelfService.mockClear();
    getGenresBookshelfService.mockClear();
  });

  it('can set booklist book isLoading', () => {
    const action = actions.getBookshelfIsLoading(LOADING_STATUSES.initial);
    expect(action).toEqual({
      type: types.FETCH_BOOKSHELF_IS_LOADING,
      isLoading: LOADING_STATUSES.initial,
    });
  });

  it('can set booklist success', () => {
    const action = actions.getBookshelfSuccess(booklist);
    expect(action).toEqual({
      type: types.FETCH_BOOKSHELF_SUCCESS,
      bookshelf: booklist,
    });
  });

  it('can save combined books success', () => {
    const action = actions.saveCombinedBooksSuccess(booklist);
    expect(action).toEqual({
      type: types.SAVE_COMBINED_BOOKS_SUCCESS,
      booklist,
    });
  });

  it('can save combined books', () => {
    const dispatch = jest.fn();
    actions.saveCombinedBooks(booklist)(dispatch);
    expect(dispatch.mock.calls).toMatchSnapshot();
  });

  it('can set filters', () => {
    const filters = [
      { key: 'unread', value: true },
      { key: 'owned', value: false },
    ];
    const actionTrue = actions.filterBookshelf(filters);
    expect(actionTrue).toEqual({
      type: types.FILTER_BOOKSHELF_SUCCESS,
      filters,
    });
  });
  it('can set booklist hasErrored', () => {
    const actionTrue = actions.getBookshelfFailure(true, 'err');
    expect(actionTrue).toEqual({
      type: types.FETCH_BOOKSHELF_HAS_ERRORED,
      hasErrored: true,
      error: 'err',
    });

    const actionFalse = actions.getBookshelfFailure(false, null);
    expect(actionFalse).toEqual({
      type: types.FETCH_BOOKSHELF_HAS_ERRORED,
      hasErrored: false,
      error: null,
    });
  });

  it('can set saveModifiedBooksSuccess', () => {
    let modifiedBooklist = Array.from(booklist);
    modifiedBooklist.differences = [
      { key: 'goodreadsRatingsCount', currentValue: 21543, newValue: 21560 },
    ];
    const action = actions.saveModifiedBooksSuccess(modifiedBooklist);
    expect(action).toEqual({
      type: types.SAVE_MODIFIED_BOOKS_SUCCESS,
      modifiedBooklist,
    });
  });

  it('can saveModifiedBooks', () => {
    let modifiedBooklist = Array.from(booklist);
    modifiedBooklist.differences = [
      { key: 'goodreadsRatingsCount', currentValue: 21543, newValue: 21560 },
    ];
    const dispatch = jest.fn();
    actions.saveModifiedBooks(modifiedBooklist)(dispatch);
    expect(dispatch.mock.calls).toMatchSnapshot();
  });

  it('insertModifiedBookSuccess', () => {
    const action = actions.insertModifiedBookSuccess();
    expect(action).toEqual({
      type: types.INSERT_MODIFIED_BOOK_SUCCESS,
    });
  });
  it('insertModifiedBook', () => {
    const dispatch = jest.fn();
    actions.insertModifiedBook(booklist[0])(dispatch);
    expect(dispatch.mock.calls).toMatchSnapshot();
  });
  describe('getBookshelf', () => {
    it('includedGenres = false, no genres selected', () => {
      const dispatch = jest.fn();
      actions.getBookshelf(false)(dispatch);
      expect(dispatch.mock.calls).toMatchSnapshot();
    });
    it('bookshelf returned with includedGenres', () => {
      const store = createMockStore();
      const bookshelf = [
        {
          categories: ['Religion'],
          read: false,
          owned: false,
          _id: '5c959505494f5dd029aeff74',
          amazonAverageRating: 4.6,
          amazonRatingsCount: 29,
          price: '',
          isbn: '9780736917728',
          title: "Raising a Daughter After God's Own Heart",
          subtitle: '',
          description:
            'Elizabeth George, bestselling author and mother of two daughters',
          thumbnail: 'http://books.google.com/books/content?id=K_AhmQEACAAJ&pr',
          goodreadsAverageRating: 4.25,
          goodreadsRatingsCount: 149,
          __v: 0,
          adjustedRating: 4.425,
        },
      ];
      const expectedResponse = [
        { isLoading: true, type: 'FETCH_BOOKSHELF_IS_LOADING' },
        { selectedGenres: ['Parenting'], type: 'SELECTED_GENRES' },
        { isLoading: false, type: 'FETCH_BOOKSHELF_IS_LOADING' },
        {
          bookshelf: [
            {
              __v: 0,
              _id: '5c959505494f5dd029aeff74',
              adjustedRating: 4.425,
              amazonAverageRating: 4.6,
              amazonRatingsCount: 29,
              categories: ['Religion'],
              description:
                'Elizabeth George, bestselling author and mother of two daughters',
              goodreadsAverageRating: 4.25,
              goodreadsRatingsCount: 149,
              isbn: '9780736917728',
              owned: false,
              price: '',
              read: false,
              subtitle: '',
              thumbnail:
                'http://books.google.com/books/content?id=K_AhmQEACAAJ&pr',
              title: "Raising a Daughter After God's Own Heart",
            },
          ],
          type: 'FETCH_BOOKSHELF_SUCCESS',
        },
        { error: null, hasErrored: false, type: 'FETCH_BOOKSHELF_HAS_ERRORED' },
        { genres: ['Religion'], type: 'FETCH_BOOKSHELF_GENRES_SUCCESS' },
      ];

      getBookshelfService.mockResolvedValue(bookshelf);
      getGenresBookshelfService.mockResolvedValue(bookshelf[0].categories);

      return store.dispatch(actions.getBookshelf(['Parenting'])).then(() => {
        expect(store.getActions()).toEqual(expectedResponse);
      });
    });
    it('getBookshelfService failed', () => {
      const store = createMockStore();
      const expectedResponse = [
        { isLoading: true, type: types.FETCH_BOOKSHELF_IS_LOADING },
        { isLoading: false, type: types.FETCH_BOOKSHELF_IS_LOADING },
        {
          error: 'err',
          hasErrored: true,
          type: types.FETCH_BOOKSHELF_HAS_ERRORED,
        },
      ];
      getBookshelfService.mockRejectedValue('err');

      return store.dispatch(actions.getBookshelf(['Parenting'])).then(() => {
        expect(store.getActions()).toEqual(expectedResponse);
      });
    });
  });

  it('addBookToBookshelfFailure', () => {
    const action = actions.addBookToBookshelfFailure('err');
    expect(action).toEqual({
      type: types.ADD_BOOK_TO_BOOKSHELF_FAILURE,
      error: 'err',
    });
  });
  it('addBookToBookshelfSuccess', () => {
    const action = actions.addBookToBookshelfSuccess(booklist);
    expect(action).toEqual({
      booklist,
      type: types.ADD_BOOK_TO_BOOKSHELF_SUCCESS,
    });
  });
  describe('addBookToBookshelf', () => {
    it('successful', () => {
      const bookshelf = [
        {
          categories: ['Juvenile Fiction'],
          read: false,
          owned: false,
          _id: '5c9d6b89f2ae7a3690d6fa42',
          amazonAverageRating: 4.7,
          amazonRatingsCount: 1916,
          price: '',
          isbn: '9780064404990',
          title: 'Year of the Griffin',
          subtitle: '',
          description:
            'It is eight years after the tours from offworld have stopped.',
          thumbnail: 'http://books.google.com/books/content?id=zCe6gRHonZg',
          goodreadsAverageRating: 4.21,
          goodreadsRatingsCount: 1860678,
          __v: 0,
          adjustedRating: 4.449916512838346,
        },
        booklist[0],
      ];
      const store = createMockStore();
      const expectedResponse = [
        { booklist: [], type: 'ADD_BOOK_TO_BOOKSHELF_SUCCESS' },
        { isLoading: true, type: 'FETCH_BOOKSHELF_IS_LOADING' },
        { selectedGenres: [], type: 'SELECTED_GENRES' },
        { isLoading: false, type: 'FETCH_BOOKSHELF_IS_LOADING' },
        {
          bookshelf: [
            {
              __v: 0,
              _id: '5c9d6b89f2ae7a3690d6fa42',
              adjustedRating: 4.449916512838346,
              amazonAverageRating: 4.7,
              amazonRatingsCount: 1916,
              categories: ['Juvenile Fiction'],
              description:
                'It is eight years after the tours from offworld have stopped.',
              goodreadsAverageRating: 4.21,
              goodreadsRatingsCount: 1860678,
              isbn: '9780064404990',
              owned: false,
              price: '',
              read: false,
              subtitle: '',
              thumbnail: 'http://books.google.com/books/content?id=zCe6gRHonZg',
              title: 'Year of the Griffin',
            },
            {
              amazonAverageRating: 4.6,
              amazonRatingsCount: 29,
              categories: ['Religion'],
              description:
                'Elizabeth George, bestselling author and mother of two daughters',
              goodreadsAverageRating: 4.25,
              goodreadsRatingsCount: 149,
              isbn: '9780736917728',
              price: '',
              subtitle: '',
              thumbnail:
                'http://books.google.com/books/content?id=K_AhmQEACAAJ&printsec=frontcover',
              title: "Raising a Daughter After God's Own Heart",
            },
          ],
          type: 'FETCH_BOOKSHELF_SUCCESS',
        },
        { error: null, hasErrored: false, type: 'FETCH_BOOKSHELF_HAS_ERRORED' },
        {
          genres: ['Juvenile Fiction', 'Religion'],
          type: 'FETCH_BOOKSHELF_GENRES_SUCCESS',
        },
      ];
      addBookshelfService.mockResolvedValue(booklist);
      getBookshelfService.mockResolvedValue(bookshelf);
      getGenresBookshelfService.mockResolvedValue([
        bookshelf[0].categories[0],
        bookshelf[1].categories[0],
      ]);

      return store.dispatch(actions.addBookToBookshelf(booklist)).then(() => {
        expect(store.getActions()).toEqual(expectedResponse);
      });
    });
    it('failed', () => {
      const store = createMockStore();
      const expectedResponse = [
        { error: 'err', type: types.ADD_BOOK_TO_BOOKSHELF_FAILURE },
      ];
      addBookshelfService.mockRejectedValue('err');

      return store.dispatch(actions.addBookToBookshelf(booklist)).then(() => {
        expect(store.getActions()).toEqual(expectedResponse);
      });
    });
  });
  it('updateBookOnBookshelfSuccess', () => {
    const action = actions.updateBookOnBookshelfSuccess();
    expect(action).toEqual({
      type: types.UPDATE_BOOK_ON_BOOKSHELF_SUCCESS,
    });
  });
  it('updateBookOnBookshelfFailure', () => {
    const action = actions.updateBookOnBookshelfFailure('err');
    expect(action).toEqual({
      type: types.UPDATE_BOOK_ON_BOOKSHELF_FAILURE,
      error: 'err',
    });
  });
  describe('updateBookOnBookshelf', () => {
    it('successful without categories field', () => {
      const store = createMockStore();
      const expectedResponse = [
        { type: types.UPDATE_BOOK_ON_BOOKSHELF_SUCCESS },
      ];
      updateBookOnBookshelfService.mockResolvedValue({
        n: 1,
        nModified: 1,
        ok: 1,
      });

      return store
        .dispatch(
          actions.updateBookOnBookshelf('5c967fbcbfb8c4eb4b9d463c', {
            goodreadsRatingsCount: 21560,
          })
        )
        .then(() => {
          expect(store.getActions()).toEqual(expectedResponse);
        });
    });
    it('successful with categories field', () => {
      const store = createMockStore();
      const expectedResponse = [
        { type: 'UPDATE_BOOK_ON_BOOKSHELF_SUCCESS' },
        { isLoading: true, type: 'FETCH_BOOKSHELF_IS_LOADING' },
        { selectedGenres: [], type: 'SELECTED_GENRES' },
        { isLoading: false, type: 'FETCH_BOOKSHELF_IS_LOADING' },
        {
          bookshelf: [
            {
              __v: 0,
              _id: '5c9d6b89f2ae7a3690d6fa42',
              adjustedRating: 4.449916512838346,
              amazonAverageRating: 4.7,
              amazonRatingsCount: 1916,
              categories: ['Juvenile Fiction'],
              description:
                'It is eight years after the tours from offworld have stopped.',
              goodreadsAverageRating: 4.21,
              goodreadsRatingsCount: 1860678,
              isbn: '9780064404990',
              owned: false,
              price: '',
              read: false,
              subtitle: '',
              thumbnail: 'http://books.google.com/books/content?id=zCe6gRHonZg',
              title: 'Year of the Griffin',
            },
            {
              amazonAverageRating: 4.6,
              amazonRatingsCount: 29,
              categories: ['Religion'],
              description:
                'Elizabeth George, bestselling author and mother of two daughters',
              goodreadsAverageRating: 4.25,
              goodreadsRatingsCount: 149,
              isbn: '9780736917728',
              price: '',
              subtitle: '',
              thumbnail:
                'http://books.google.com/books/content?id=K_AhmQEACAAJ&printsec=frontcover',
              title: "Raising a Daughter After God's Own Heart",
            },
          ],
          type: 'FETCH_BOOKSHELF_SUCCESS',
        },
        { error: null, hasErrored: false, type: 'FETCH_BOOKSHELF_HAS_ERRORED' },
        {
          genres: ['Juvenile Fiction', 'Religion'],
          type: 'FETCH_BOOKSHELF_GENRES_SUCCESS',
        },
      ];

      updateBookOnBookshelfService.mockResolvedValue({
        n: 1,
        nModified: 1,
        ok: 1,
      });

      return store
        .dispatch(
          actions.updateBookOnBookshelf('5c967fbcbfb8c4eb4b9d463c', {
            goodreadsRatingsCount: 21560,
            categories: ['Juvenile Fiction'],
          })
        )
        .then(() => {
          expect(store.getActions()).toEqual(expectedResponse);
        });
    });
    it('calls getBookshelf on refreshBookshelf', () => {
      const store = createMockStore();
      const expectedResponse = [{ type: 'UPDATE_BOOK_ON_BOOKSHELF_SUCCESS' }];
      updateBookOnBookshelfService.mockResolvedValue({
        n: 1,
        nModified: 1,
        ok: 1,
      });

      return store
        .dispatch(
          actions.updateBookOnBookshelf(
            '5c967fbcbfb8c4eb4b9d463c',
            {
              goodreadsRatingsCount: 21560,
            },
            true
          )
        )
        .then(() => {
          expect(store.getActions()).toEqual(expectedResponse);
        });
    });
    it('fails', () => {
      const store = createMockStore();
      const expectedResponse = [
        { error: 'err', type: types.UPDATE_BOOK_ON_BOOKSHELF_FAILURE },
      ];
      updateBookOnBookshelfService.mockRejectedValue(expectedResponse[0].error);

      return store
        .dispatch(
          actions.updateBookOnBookshelf('5c967fbcbfb8c4eb4b9d463c', {
            goodreadsRatingsCount: 21560,
          })
        )
        .then(() => {
          expect(store.getActions()).toEqual(expectedResponse);
        });
    });
  });
  it('deleteBookOnBookshelfSuccess', () => {
    const action = actions.deleteBookOnBookshelfSuccess(123);
    expect(action).toEqual({
      type: types.DELETE_BOOK_ON_BOOKSHELF_SUCCESS,
      deleteId: 123,
    });
  });
  it('deleteBookOnBookshelfFailure', () => {
    const action = actions.deleteBookOnBookshelfFailure('err');
    expect(action).toEqual({
      type: types.DELETE_BOOK_ON_BOOKSHELF_FAILURE,
      error: 'err',
    });
  });
  describe('deleteBookOnBookshelf', () => {
    it('successful', () => {
      const store = createMockStore();
      const expectedResponse = [
        {
          deleteId: '5c967fbcbfb8c4eb4b9d463c',
          type: 'DELETE_BOOK_ON_BOOKSHELF_SUCCESS',
        },
        { isLoading: true, type: 'FETCH_BOOKSHELF_IS_LOADING' },
        { selectedGenres: [], type: 'SELECTED_GENRES' },
        { isLoading: false, type: 'FETCH_BOOKSHELF_IS_LOADING' },
        {
          bookshelf: [
            {
              __v: 0,
              _id: '5c9d6b89f2ae7a3690d6fa42',
              adjustedRating: 4.449916512838346,
              amazonAverageRating: 4.7,
              amazonRatingsCount: 1916,
              categories: ['Juvenile Fiction'],
              description:
                'It is eight years after the tours from offworld have stopped.',
              goodreadsAverageRating: 4.21,
              goodreadsRatingsCount: 1860678,
              isbn: '9780064404990',
              owned: false,
              price: '',
              read: false,
              subtitle: '',
              thumbnail: 'http://books.google.com/books/content?id=zCe6gRHonZg',
              title: 'Year of the Griffin',
            },
            {
              amazonAverageRating: 4.6,
              amazonRatingsCount: 29,
              categories: ['Religion'],
              description:
                'Elizabeth George, bestselling author and mother of two daughters',
              goodreadsAverageRating: 4.25,
              goodreadsRatingsCount: 149,
              isbn: '9780736917728',
              price: '',
              subtitle: '',
              thumbnail:
                'http://books.google.com/books/content?id=K_AhmQEACAAJ&printsec=frontcover',
              title: "Raising a Daughter After God's Own Heart",
            },
          ],
          type: 'FETCH_BOOKSHELF_SUCCESS',
        },
        { error: null, hasErrored: false, type: 'FETCH_BOOKSHELF_HAS_ERRORED' },
        { genres: ['Religion'], type: 'FETCH_BOOKSHELF_GENRES_SUCCESS' },
      ];
      deleteBookOnBookshelfService.mockResolvedValue(booklist[0]);
      getGenresBookshelfService.mockResolvedValue(booklist[0].categories);

      return store
        .dispatch(actions.deleteBookOnBookshelf('5c967fbcbfb8c4eb4b9d463c'))
        .then(() => {
          expect(store.getActions()).toEqual(expectedResponse);
        });
    });
    it('failed', () => {
      const store = createMockStore();
      const expectedResponse = [
        { error: 'err', type: types.DELETE_BOOK_ON_BOOKSHELF_FAILURE },
      ];
      deleteBookOnBookshelfService.mockRejectedValue(expectedResponse[0].error);

      return store
        .dispatch(actions.deleteBookOnBookshelf('5c967fbcbfb8c4eb4b9d463c'))
        .then(() => {
          expect(store.getActions()).toEqual(expectedResponse);
        });
    });
  });
  describe('getBookshelfGenres', () => {
    it('successful', () => {
      const store = createMockStore();
      const expectedResponse = [
        {
          genres: ['Genre1', 'Genre2'],
          type: types.FETCH_BOOKSHELF_GENRES_SUCCESS,
        },
      ];
      getGenresBookshelfService.mockResolvedValue(['Genre1', 'Genre2']);

      return store.dispatch(actions.getBookshelfGenres()).then(() => {
        expect(store.getActions()).toEqual(expectedResponse);
      });
    });
    it('failed', () => {
      const store = createMockStore();
      const expectedResponse = [
        { error: 'err', type: types.FETCH_BOOKSHELF_GENRES_FAILURE },
      ];
      getGenresBookshelfService.mockRejectedValue(expectedResponse[0].error);

      return store.dispatch(actions.getBookshelfGenres()).then(() => {
        expect(store.getActions()).toEqual(expectedResponse);
      });
    });
  });
  it('deleteModifiedBookFromBooklist', () => {
    const action = actions.deleteModifiedBookFromBooklist('12345');
    expect(action).toEqual({
      type: types.DELETE_BOOK_FROM_BOOKLIST_SUCCESS,
      deleteISBN: '12345',
    });
  });
  it('deleteModifiedBook', () => {
    const dispatch = jest.fn();
    actions.deleteModifiedBook('12345')(dispatch);
    expect(dispatch.mock.calls).toMatchSnapshot();
  });

  it('getBookshelfGenresFailure', () => {
    const action = actions.getBookshelfGenresFailure('err');
    expect(action).toEqual({
      type: types.FETCH_BOOKSHELF_GENRES_FAILURE,
      error: 'err',
    });
  });
  it('can clearBooks', () => {
    const dispatch = jest.fn();
    actions.clearBooks()(dispatch);
    expect(dispatch.mock.calls).toMatchSnapshot();
  });
});
