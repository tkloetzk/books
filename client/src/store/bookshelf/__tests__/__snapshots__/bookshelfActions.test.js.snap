// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bookshelf actions can clearBooks 1`] = `
Array [
  Array [
    [Function],
  ],
  Array [
    [Function],
  ],
  Array [
    [Function],
  ],
]
`;

exports[`bookshelf actions can save combined books 1`] = `
Array [
  Array [
    Object {
      "booklist": Array [
        Object {
          "amazonAverageRating": 4.6,
          "amazonRatingsCount": 29,
          "categories": Array [
            "Religion",
          ],
          "description": "Elizabeth George, bestselling author and mother of two daughters",
          "goodreadsAverageRating": 4.25,
          "goodreadsRatingsCount": 149,
          "isbn": "9780736917728",
          "price": "",
          "subtitle": "",
          "thumbnail": "http://books.google.com/books/content?id=K_AhmQEACAAJ&printsec=frontcover",
          "title": "Raising a Daughter After God's Own Heart",
        },
      ],
      "type": "SAVE_COMBINED_BOOKS_SUCCESS",
    },
  ],
]
`;

exports[`bookshelf actions can saveModifiedBooks 1`] = `
Array [
  Array [
    Object {
      "modifiedBooklist": Array [
        Object {
          "amazonAverageRating": 4.6,
          "amazonRatingsCount": 29,
          "categories": Array [
            "Religion",
          ],
          "description": "Elizabeth George, bestselling author and mother of two daughters",
          "goodreadsAverageRating": 4.25,
          "goodreadsRatingsCount": 149,
          "isbn": "9780736917728",
          "price": "",
          "subtitle": "",
          "thumbnail": "http://books.google.com/books/content?id=K_AhmQEACAAJ&printsec=frontcover",
          "title": "Raising a Daughter After God's Own Heart",
        },
      ],
      "type": "SAVE_MODIFIED_BOOKS_SUCCESS",
    },
  ],
]
`;
